lib/eventbus.js

```diff
@@ -0,0 +1,159 @@
+'use strict';
+
+const _        = require('lodash');
+const P        = require('bluebird');
+const bunyan          = require('bunyan');
+
+const EventInvalidError = require('../lib/validator').EventInvalidError;
+
+
+/**
+ * Represents and Eventbus process status result.
+ * Each event passed to Eventbus process() will map
+ * to an final EventStatus.  If the status is a failure
+ * this.context should contain the error.
+ */
+class EventStatus {
+    constructor(status, context, event) {
+        this.status = status;
+        this.context = context;
+        this.event = event;
+    }
+
+    /**
+     * static 'enum' constant for possible status types.
+     */
+    static STATUSES() {
+        return [
+            'success',
+            'error',
+            'invalid'
+        ];
+    }
+}
+
+/**
+ * An Eventbus asynchronously validates and produces a list of events.
+ * The validate function and produce function implementations are passed
+ * in by the user.  It is expected that validate return a Promise of
+ * the validated event or throws an EventInvalidError, and that produce returns a
+ * Promise of a produced event result or throws an Error.
+ *
+ * Once finished, events will be returned grouped by their status.
+ */
+class Eventbus {
+    /**
+     * @param {function<(Object) => Object>} validate (event) => event
+     * @param {function<(Object) => Object>} produce (event) => result
```

Comment:
Would it be better to pass the whole validator/producer object with a known interface?